pipeline/backend/kubernetes/kubernetes.go

```diff
@@ -231,43 +233,34 @@ func (e *kube) WaitStep(ctx context.Context, step *types.Step, taskUUID string)
 			return
 		}
 
-		if pod.Name == podName {
-			if isImagePullBackOffState(pod) {
-				finished <- true
-			}
-			if isCompleted(pod) {
-				finished <- true
-			}
+		if pod.Name != podName {
+			return
+		}
 
-			switch pod.Status.Phase {
-			case v1.PodSucceeded, v1.PodFailed, v1.PodUnknown:
-				finished <- true
-			}
+		if isImagePullBackOffState(pod) || isCompleted(pod) || pod.Status.Phase == v1.PodSucceeded || pod.Status.Phase == v1.PodFailed || pod.Status.Phase == v1.PodUnknown {
+			finished <- true
 		}
 	}
 
-	// TODO 5 seconds is against best practice, k3s didn't work otherwise
 	si := informers.NewSharedInformerFactoryWithOptions(e.client, 5*time.Second, informers.WithNamespace(e.config.Namespace))
-	if _, err := si.Core().V1().Pods().Informer().AddEventHandler(
-		cache.ResourceEventHandlerFuncs{
-			UpdateFunc: podUpdated,
-		},
-	); err != nil {
+	podInformer := si.Core().V1().Pods().Informer()
+
+	if _, err := podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: podUpdated,
+	}); err != nil {
 		return nil, err
 	}
 
 	stop := make(chan struct{})
-	si.Start(stop)
 	defer close(stop)
 
-	// TODO Cancel on ctx.Done
-	<-finished
+	go podInformer.Run(stop)
```

Comment:
What's the difference between podInformer.Run and si.Start ? Is this what fixes the problem?